plugins/ua_nodestore_binary_encoded.c

```diff
@@ -0,0 +1,1163 @@
+        return ZIP_CMP_LESS;
+    if(aa->nodeIdHash > bb->nodeIdHash)
+        return ZIP_CMP_MORE;
+
+    /* Compore nodes in detail */
+    return (enum ZIP_CMP)UA_NodeId_order(&aa->nodeId, &bb->nodeId);
+}
+
+ZIP_HEAD(NodeTreeBin, NodeEntry);
+typedef struct NodeTreeBin NodeTreeBin;
+
+typedef struct {
+    NodeTreeBin root;
+} ZipContext;
+
+ZIP_PROTTYPE(NodeTreeBin, NodeEntry, NodeEntry)
+ZIP_IMPL(NodeTreeBin, NodeEntry, zipfields, NodeEntry, zipfields, cmpNodeId)
+
+static NodeEntry *
+newEntry(UA_NodeClass nodeClass) {
+    size_t size = sizeof(NodeEntry) - sizeof(UA_NodeId);
+    switch(nodeClass) {
+    case UA_NODECLASS_OBJECT:
+        size += sizeof(UA_ObjectNode);
+        break;
+    case UA_NODECLASS_VARIABLE:
+        size += sizeof(UA_VariableNode);
+        break;
+    case UA_NODECLASS_METHOD:
+        size += sizeof(UA_MethodNode);
+        break;
+    case UA_NODECLASS_OBJECTTYPE:
+        size += sizeof(UA_ObjectTypeNode);
+        break;
+    case UA_NODECLASS_VARIABLETYPE:
+        size += sizeof(UA_VariableTypeNode);
+        break;
+    case UA_NODECLASS_REFERENCETYPE:
+        size += sizeof(UA_ReferenceTypeNode);
+        break;
+    case UA_NODECLASS_DATATYPE:
+        size += sizeof(UA_DataTypeNode);
+        break;
+    case UA_NODECLASS_VIEW:
+        size += sizeof(UA_ViewNode);
+        break;
+    default:
+        return NULL;
+    }
+    NodeEntry *entry = (NodeEntry*)UA_calloc(1, size);
+    if(!entry)
+        return NULL;
+    UA_Node *node = (UA_Node*)&entry->nodeId;
+    node->nodeClass = nodeClass;
+    return entry;
+}
+
+static void
+deleteEntry(NodeEntry *entry) {
+    UA_Node_clear((UA_Node*)&entry->nodeId);
+    UA_free(entry);
+}
+
+static void
+cleanupEntry(NodeEntry *entry) {
+    if(entry->deleted && entry->refCount == 0)
+        deleteEntry(entry);
+}
+
+/***********************/
+/* Interface functions */
+/***********************/
+
+/* Not yet inserted into the ZipContext */
+static UA_Node *
+zipNsNewNode(void *nsCtx, UA_NodeClass nodeClass) {
+    NodeEntry *entry = newEntry(nodeClass);
+    if(!entry)
+        return NULL;
+    return (UA_Node*)&entry->nodeId;
+}
+
+/* Not yet inserted into the ZipContext */
+static void
+zipNsDeleteNode(void *nsCtx, UA_Node *node) {
+    deleteEntry(container_of(node, NodeEntry, nodeId));
+}
+
+static void
+zipNsReleaseNode(void *nsCtx, const UA_Node *node) {
+    if(!node)
+        return;
+    NodeEntry *entry = container_of(node, NodeEntry, nodeId);
+    UA_assert(entry->refCount > 0);
+    --entry->refCount;
+    if(entry->deleted) {
+        /* Encode the edited node and replace the encoded content */
+        if(node->nodeClass == UA_NODECLASS_VARIABLE) {
+            encodeEditedNode(node, &encodeBin, ltRead, ltSizeRead);
+        }
+    }
+    cleanupEntry(entry);
+}
+
+UA_Node*
+decodeNode(void *ctx, UA_ByteString encodedBin, size_t offset) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_NodeId nodeID;
```

Comment:
If you malloc the new node first, then you can directly decode "into" the node.
Instead of allocating on the stack and then doing a memcpy.